pyk/src/pyk/kore/rule.py

```diff
@@ -0,0 +1,287 @@
+"""Parse KORE axioms into rewrite rules.
+
+Based on the [LLVM Backend's implementation](https://github.com/runtimeverification/llvm-backend/blob/d5eab4b0f0e610bc60843ebb482f79c043b92702/lib/ast/pattern_matching.cpp).
+"""
+
+from __future__ import annotations
+
+from abc import ABC
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, final
+
+from .syntax import And, App, Ceil, Equals, EVar, Implies, In, Not, Rewrites, String, Top
+
+if TYPE_CHECKING:
+    from .syntax import Axiom, Definition, Pattern
+
+    Attrs = dict[str, tuple[Pattern, ...]]
+
+
+class Rule(ABC):
+    lhs: Pattern
+    rhs: Pattern
+    req: Pattern | None
+    ens: Pattern | None
+    priority: int
+
+    @staticmethod
+    def from_axiom(axiom: Axiom) -> Rule:
+        if isinstance(axiom.pattern, Rewrites):
+            return RewriteRule.from_axiom(axiom)
+
+        attrs = (attr.symbol for attr in axiom.attrs)
+        if 'simplification' in attrs:
+            return SimpliRule.from_axiom(axiom)
+
+        return FunctionRule.from_axiom(axiom)
+
+    @staticmethod
+    def extract_all(defn: Definition) -> list[Rule]:
+        filter_attrs = [
+            'assoc',
+            'constructor',
+            'functional',
+            'idem',
+            'symbol-overload',
+            'subsort',
+            'unit',
+        ]
+
+        res: list[Rule] = []
+        for axiom in defn.axioms:
+            attrs = {attr.symbol for attr in axiom.attrs}
+            if any(attr in attrs for attr in filter_attrs):
+                continue
+
+            match axiom.pattern:
+                case Equals():
+                    # A single simplification from prelude module INJ:
+                    # axiom {S1, S2, S3, R} \equals{S3, R}(inj{S2, S3}(inj{S1, S2}(T:S1)), inj{S1, S3}(T:S1)) [simplification{}()]
+                    continue
```

Comment:
Maybe a good idea to double-check that it is indeed this particular simplification equation? (by checking the symbol on the LHS and maybe verifying that it does not have location information)

When developing booster, we tried to be specific about the cases that we discard and not have a catch-all case so we would get to know if the classification is incomplete.

Reply:
Done.